### ISSUE TITLE
Connection to bus lost after time

### ISSUE THREAD

**Issue by JornHermans** [velbus/velbus-tcp-snap#3]
I have 2 cases where the connection to the Velbus BUS is lost, the snap is still running and connections are still possible.
After doing a `snap restart velbus-tcp` everything is working again.

**Comment by silamon** [comment]
Do you have any extra logging? That would help us to find the problem better. 
You can enable this through configuration. `snap set velbus-tcp logging.type=debug`.

I'll look into how stable 1.3.0/1.4.0 is, as I'm unsure if it was tested or not.

**Comment by silamon** [comment]
If you want, you can send us a log privately. I'll send you a mail.

**Comment by JornHermans** [comment]
I have no extra logging, i wil enable this. 
And upload what the result is.

**Comment by JornHermans** [comment]
When doing this the snap is also restarted, with as result that the connection is restored. 

Current Logging:

> 2020-11-06T20:42:33Z systemd[1]: snap.velbus-tcp.velbus-tcp.service: Succeeded.
> 2020-11-06T20:42:33Z systemd[1]: Stopped Service for snap application velbus-tcp.velbus-tcp.
> 2020-11-06T20:42:33Z systemd[1]: Started Service for snap application velbus-tcp.velbus-tcp.
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,592 - INFO - bus - Autodiscovered 1 port(s): ['/dev/ttyACM0']
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,594 - INFO - bus - Choosing /dev/ttyACM0
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,596 - INFO - bus - Serial connection active on port /dev/ttyACM0
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,596 - DEBUG - bridge - [BUS IN] 0xf 0xfb 0xcc 0x7 0xe6 0x33 0xa0 0x31 0x0 0x3a 0xa0 0x5f 0x4
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,601 - INFO - network - Listening to TCP connections on 0.0.0.0:27015 [SSL:True]
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,602 - INFO - network - Listening to TCP connections on 127.0.0.1:3788 [SSL:False]
> 2020-11-06T20:42:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-06 21:42:34,670 - INFO - network - TCP connection from ('127.0.0.1', 34354)

When i notice that the connection is dropped again and a restart of the snap is required i will upload the log again.

**Comment by JornHermans** [comment]
I've got this from logging, sometimes its possible to get a command to the bus (takes a lot of time).
It's almost failing in this case...

> 2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,215 - INFO - network - TCP connection from ('127.0.0.1', 54124)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,216 - INFO - network - TCP connection closed ('127.0.0.1', 54124)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,225 - INFO - network - TCP connection from ('127.0.0.1', 54128)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,226 - INFO - network - TCP connection closed ('127.0.0.1', 54128)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,243 - ERROR - client - [Errno 104] Connection reset by peer
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: Traceback (most recent call last):
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]:   File "/snap/velbus-tcp/105/lib/connection/client.py", line 117, in __recv
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]:     data = self.__connection.recv(1024)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: ConnectionResetError: [Errno 104] Connection reset by peer
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: Exception in thread Receive from client thread:
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: Traceback (most recent call last):
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:   File "/snap/velbus-tcp/105/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:     self.run()
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:   File "/snap/velbus-tcp/105/usr/lib/python3.6/threading.py", line 864, in run
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:     self._target(*self._args, **self._kwargs)
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:   File "/snap/velbus-tcp/105/lib/connection/client.py", line 134, in __recv
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:     self.stop()
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:   File "/snap/velbus-tcp/105/lib/connection/client.py", line 49, in stop
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]:     self.__connection.shutdown(socket.SHUT_RDWR)
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: OSError: [Errno 107] Transport endpoint is not connected
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:31,062 - DEBUG - bridge - [BUS IN] 0xf 0xf8 0x5d 0x4 0x0 0x10 0x0 0x0 0x88 0x4
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:31,063 - DEBUG - network - [TCP OUT] 0xf 0xf8 0x5d 0x4 0x0 0x10 0x0 0x0 0x88 0x4
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:31,792 - DEBUG - bridge - [BUS IN] 0xf 0xfb 0x5d 0x2 0xf6 0x10 0x91 0x4
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:31,792 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5d 0x2 0xf6 0x10 0x91 0x4
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:31,792 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5d 0x2 0xf6 0x10 0x91 0x4
2020-11-09T10:16:32Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:32,456 - DEBUG - bridge - [BUS IN] 0xf 0xfb 0x5c 0x8 0xfb 0x10 0x0 0x1 0x80 0x0 0x0 0x0 0x6 0x4
2020-11-09T10:16:32Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:32,458 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5c 0x8 0xfb 0x10 0x0 0x1 0x80 0x0 0x0 0x0 0x6 0x4
2020-11-09T10:16:32Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:32,459 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5c 0x8 0xfb 0x10 0x0 0x1 0x80 0x0 0x0 0x0 0x6 0x4
2020-11-09T10:16:33Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:33,085 - DEBUG - bridge - [BUS IN] 0xf 0xf8 0x5c 0x4 0x0 0x10 0x0 0x0 0x89 0x4
2020-11-09T10:16:33Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:33,085 - DEBUG - network - [TCP OUT] 0xf 0xf8 0x5c 0x4 0x0 0x10 0x0 0x0 0x89 0x4
2020-11-09T10:16:33Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:33,730 - DEBUG - bridge - [BUS IN] 0xf 0xf8 0x5d 0x4 0x0 0x0 0x10 0x0 0x88 0x4
2020-11-09T10:16:33Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:33,730 - DEBUG - network - [TCP OUT] 0xf 0xf8 0x5d 0x4 0x0 0x0 0x10 0x0 0x88 0x4
2020-11-09T10:16:33Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:33,731 - DEBUG - network - [TCP OUT] 0xf 0xf8 0x5d 0x4 0x0 0x0 0x10 0x0 0x88 0x4
2020-11-09T10:16:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:34,386 - DEBUG - bridge - [BUS IN] 0xf 0xfb 0x5d 0x2 0xf5 0x10 0x92 0x4
2020-11-09T10:16:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:34,386 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5d 0x2 0xf5 0x10 0x92 0x4
2020-11-09T10:16:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:34,387 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5d 0x2 0xf5 0x10 0x92 0x4
2020-11-09T10:16:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:34,964 - DEBUG - bridge - [BUS IN] 0xf 0xfb 0x5c 0x8 0xfb 0x10 0x0 0x0 0x0 0x0 0x0 0x0 0x87 0x4
2020-11-09T10:16:34Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:34,965 - DEBUG - network - [TCP OUT] 0xf 0xfb 0x5c 0x8 0xfb 0x10 0x0 0x0 0x0 0x0 0x0 0x0 0x87 0x4
2020-11-09T10:16:35Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:35,594 - DEBUG - bridge - [BUS IN] 0xf 0xf8 0x5c 0x4 0x0 0x0 0x10 0x0 0x89 0x4
2020-11-09T10:16:35Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:35,594 - DEBUG - network - [TCP OUT] 0xf 0xf8 0x5c 0x4 0x0 0x0 0x10 0x0 0x89 0x4

I hope this can help you debug the issue

**Comment by dukeofphilberg** [comment]
What's happening is rather the TCP socket is losing its client's connection, rather than bus connection being lost.

Try to setup a check when your client disconnects.

```
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,215 - INFO - network - TCP connection from ('127.0.0.1', 54124)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,216 - INFO - network - TCP connection closed ('127.0.0.1', 54124)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,225 - INFO - network - TCP connection from ('127.0.0.1', 54128)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,226 - INFO - network - TCP connection closed ('127.0.0.1', 54128)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: 2020-11-09 11:16:07,243 - ERROR - client - [Errno 104] Connection reset by peer
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: Traceback (most recent call last):
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: File "/snap/velbus-tcp/105/lib/connection/client.py", line 117, in __recv
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: data = self.__connection.recv(1024)
2020-11-09T10:16:07Z velbus-tcp.velbus-tcp[28558]: ConnectionResetError: [Errno 104] Connection reset by peer
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: Exception in thread Receive from client thread:
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: Traceback (most recent call last):
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: File "/snap/velbus-tcp/105/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: self.run()
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: File "/snap/velbus-tcp/105/usr/lib/python3.6/threading.py", line 864, in run
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: self._target(*self._args, **self._kwargs)
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: File "/snap/velbus-tcp/105/lib/connection/client.py", line 134, in __recv
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: self.stop()
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: File "/snap/velbus-tcp/105/lib/connection/client.py", line 49, in stop
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: self.__connection.shutdown(socket.SHUT_RDWR)
2020-11-09T10:16:31Z velbus-tcp.velbus-tcp[28558]: OSError: [Errno 107] Transport endpoint is not connected
```

**Comment by JornHermans** [comment]
We already did implement a permanent TCP connection on an other device. 

In the logs for this device we open a connection for each time we send a message to the velbus bus, also we have one open to monitor the velbus bus.

When i use the follow function on the snap logs, where the device doesn't get connection with the velbus bus i can see the TCP connection being opened and closed again (= normal behavior, on that device). So that part is working fine, still the message sended is not going to the velbus bus, in this case the message could be send after a few times of repeating the message, some time later on you can't send or receive anything to or from the velbus bus.

Also when we implant a hourly restart of the snap everything keeps working just fine (= bandage not a lifetime solution).
